docs/contributing/retries-and-waiters.md

```diff
@@ -0,0 +1,474 @@
+
+### Terraform AWS Provider Service Retries
+
+The AWS Go SDK middleware provides hooks for injecting custom logic in the service client handlers. This type of handling is preferred when retry behavior must be applied in many resources, such as an error code that is not marked as automatically retryable by the service API. The Terraform AWS Provider includes some additional retry changing behaviors using this method, which can be found in the `aws/config.go` file. For example:
+
+```go
+client.kafkaconn.Handlers.Retry.PushBack(func(r *request.Request) {
+	if tfaws.ErrorMessageContains(r.Error, kafka.ErrCodeTooManyRequestsException, "Too Many Requests") {
+		r.Retryable = aws.Bool(true)
+	}
+})
+```
+
+## Eventual Consistency
+
+Eventual consistency is a temporary condition where the remote system can return outdated information or errors due to not being strongly read-after-write consistent. This is a pattern found in remote systems that must be highly scaled for broad usage.
+
+Terraform expects any planned resource lifecycle change (create, update, destroy of the resource itself) and planned resource attribute value change to match after being applied. Conversely, operators typically expect that Terraform resources also implement the concept of drift detection for resources and their attributes, which requires reading information back from the remote system after an operation. A common implementation is refreshing the Terraform State information (`d.Set()`) during the `Read` function of a resource after `Create` and `Update`.
+
+These two concepts conflict with each other and require additional handling in Terraform resource logic as shown in the following sections. These issues are _not_ reliably reproducible, especially in the case of writing acceptance testing, so they can be elusive with false positives to verify fixes.
+
+### Operation Specific Error Retries
+
+Given a properly ordered Terraform configuration, eventual consistency can unexpectedly prevent downstream operations from being successful. To reduce frustrating behavior for operators where a simple retry after a few seconds would resolve the error, AWS Go SDK operations can be wrapped with the `resource.Retry()` or `resource.RetryContext()` function. These retries should have a reasonably low timeout (typically two minutes but up to five minutes) that are saved in a constant for reusability. These functions are preferably inline with the associated resource logic, to remove any indirection with the code.
+
+This type of logic should not be implemented to overcome improperly ordered Terraform configurations as it is not guaranteed to work in larger environments.
+
+```go
+// aws/internal/service/{SERVICE}/waiter/consts.go (created if does not exist)
+package {SERVICE}
+
+const (
+	// Maximum amount of time to wait for Example eventual consistency
+	ExampleTimeout = 2 * time.Minute
```

Comment:
If _Example_ is the service name and we are in a service-specific package, should this be `PropagationTimeout = 2 * time.Minute` instead (like it is for IAM)?

Reply:
Fixed in latest commit.